go/vt/vttablet/tabletserver/state_manager.go

```diff
@@ -330,11 +334,25 @@ func (sm *stateManager) checkMySQL() {
 		}
 		defer sm.transitioning.Release()
 
+		// This is required to prevent new queries from running in StartRequest
+		// unless they are part of a running transaction.
+		sm.setWantState(StateNotConnected)
 		sm.closeAll()
+
+		// Now that we reached the NotConnected state, we want to go back to the
+		// Serving state. The retry will only succeed once MySQL is reachable again
+		// Until then EnsureConnectionAndDB will error out.
+		sm.setWantState(StateServing)
 		sm.retryTransition(fmt.Sprintf("Cannot connect to MySQL, shutting down query service: %v", err))
 	}()
```

Comment:
changes are good, 

Just wanted to know why we cannot just call `SetServingType` methods which take the same actions. Also, we would have a single code path for state changes.

Reply:
We can't use that because in `mustTransition` we use `Acquire` whereas here we want to use `TryAcquire`. Also that call requires a tablet type to call to, which is something we don't want to tweek at all. So there is a trade-off between the two implementations, one. will have a single code-path but with many more if-else conditions, and this wherein there are 2 code-paths but both succinct in their implementations